polygeist/tools/cgeist/Lib/CGCall.cc

```diff
@@ -79,12 +80,82 @@ static void castCallerArgs(mlir::func::FuncOp callee,
   }
 }
 
+/// Return whether a builder is or not inserting in a device function.
+static bool inDeviceFunc(OpBuilder &Builder) {
+  auto Func =
+      Builder.getInsertionBlock()->getParent()->getParentOfType<func::FuncOp>();
+  return !Func || mlirclang::isDeviceFunction(Func);
+}
+
+template <typename OwnMapTy, typename OtherMapTy, typename ModuleTy>
+func::FuncOp cloneFunction(func::FuncOp Function, OwnMapTy &OwnMap,
+                           const OtherMapTy &OtherMap, ModuleTy &Module) {
+  const auto Name = Function.getName().str();
+  auto NewFunction = Function.clone();
+  OwnMap[Name] = NewFunction;
+  Module->push_back(NewFunction);
+  // Clone all functions called from this function.
+  NewFunction.walk([&](CallOp Call) {
+    const auto Name = Call.getCallee().str();
+    if (OwnMap.find(Name) != OwnMap.end()) {
+      return;
+    }
+    auto Func = cast<func::FuncOp>(OtherMap.find(Name)->second);
+    cloneFunction(Func, OwnMap, OtherMap, Module);
+  });
+
+  return NewFunction;
+}
+
+func::FuncOp MLIRScanner::CloneFunction(func::FuncOp Function,
+                                        bool ToGPUModule) {
+  return ToGPUModule
+             ? cloneFunction(Function, Glob.kernels, Glob.functions, gpuModule)
+             : cloneFunction(Function, Glob.functions, Glob.kernels, module);
```

Comment:
Wondering why the 2 calls to `cloneFunction` receive the `Glob.kernels` and `Glob.functions` in different order.